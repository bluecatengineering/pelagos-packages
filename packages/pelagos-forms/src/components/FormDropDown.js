import {DropDownField} from '@bluecateng/pelagos';
import {connect} from '@bluecateng/auto-forms';

export default connect(DropDownField, ({value, error, setValue}) => ({
	value,
	error,
	onChange: setValue,
}));
