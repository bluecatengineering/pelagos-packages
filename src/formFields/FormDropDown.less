.FormDropDown {
	margin-bottom: 10px;
}
