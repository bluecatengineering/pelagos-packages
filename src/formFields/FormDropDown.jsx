import React from 'react';
import PropTypes from 'prop-types';

import Label from '../components/Label';
import Select from '../components/Select';
import './FormDropDown.less';

const FormDropDown = ({
	componentId,
	className,
	label,
	value,
	options,
	placeholder,
	disabled,
	error,
	getOptionValue,
	renderOption,
	onChange,
}) => (
	<div className={'FormDropDown' + (className ? ' ' + className : '')}>
		<div>
			<Label text={label} />
		</div>
		<Select
			componentId={componentId}
			className="FormDropDown__select"
			value={value}
			options={options}
			placeholder={placeholder}
			disabled={disabled}
			error={error}
			getOptionValue={getOptionValue}
			renderOption={renderOption}
			onChange={onChange}
		/>
	</div>
);

FormDropDown.propTypes = {
	componentId: PropTypes.string,
	className: PropTypes.string,
	label: PropTypes.string.isRequired,
	value: PropTypes.string,
	options: PropTypes.array.isRequired,
	placeholder: PropTypes.string,
	disabled: PropTypes.bool,
	error: PropTypes.bool,
	getOptionValue: PropTypes.func,
	renderOption: PropTypes.func.isRequired,
	onChange: PropTypes.func.isRequired,
};

export default FormDropDown;
