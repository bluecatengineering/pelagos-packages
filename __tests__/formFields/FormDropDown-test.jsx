import React from 'react';
import {shallow} from 'enzyme';

import FormDropDown from '../../src/formFields/FormDropDown';

jest.unmock('../../src/formFields/FormDropDown');

const values = {
	o0: 'Option 0',
	o1: 'Option 1',
};
const options = Object.keys(values);
const renderOption = o => values[o];

describe('FormDropDown', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<FormDropDown
					id="test"
					componentId="test"
					label="Test"
					value="o0"
					options={options}
					renderOption={renderOption}
					onChange={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when className is set', () => {
			const wrapper = shallow(
				<FormDropDown
					componentId="id"
					className="TestClass"
					label="Test"
					value="o0"
					options={options}
					renderOption={renderOption}
					onChange={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when there are errors', () => {
			const options = [{componentId: 'id', value: 'o0', text: 'Option 0'}, {value: 'o1', text: 'Option 1'}];
			const wrapper = shallow(
				<FormDropDown
					componentId="id"
					label="Test"
					value="o0"
					options={options}
					error={true}
					renderOption={renderOption}
					onChange={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onChange when the value changes', () => {
			const onChange = jest.fn();
			const wrapper = shallow(
				<FormDropDown componentId="id" label="Test" options={[]} renderOption={renderOption} onChange={onChange} />
			);
			wrapper.find('Select').simulate('change', 'test');
			expect(onChange.mock.calls).toEqual([['test']]);
		});
	});
});
