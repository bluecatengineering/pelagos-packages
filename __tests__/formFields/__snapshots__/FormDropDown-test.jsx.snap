// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDropDown rendering renders expected elements 1`] = `
<div
  className="FormDropDown"
>
  <div>
    <Label
      text="Test"
    />
  </div>
  <Select
    className="FormDropDown__select"
    componentId="id"
    getOptionValue={[MockFunction]}
    onChange={[MockFunction]}
    options={
      Array [
        "o0",
        "o1",
      ]
    }
    renderOption={[Function]}
    value="o0"
  />
</div>
`;

exports[`FormDropDown rendering renders expected elements when className is set 1`] = `
<div
  className="FormDropDown TestClass"
>
  <div>
    <Label
      text="Test"
    />
  </div>
  <Select
    className="FormDropDown__select"
    componentId="id"
    getOptionValue={[MockFunction]}
    onChange={[MockFunction]}
    options={
      Array [
        "o0",
        "o1",
      ]
    }
    renderOption={[Function]}
    value="o0"
  />
</div>
`;

exports[`FormDropDown rendering renders expected elements when there are errors 1`] = `
<div
  className="FormDropDown"
>
  <div>
    <Label
      text="Test"
    />
  </div>
  <Select
    className="FormDropDown__select"
    componentId="id"
    error={true}
    getOptionValue={[MockFunction]}
    onChange={[MockFunction]}
    options={
      Array [
        Object {
          "componentId": "id",
          "text": "Option 0",
          "value": "o0",
        },
        Object {
          "text": "Option 1",
          "value": "o1",
        },
      ]
    }
    renderOption={[Function]}
    value="o0"
  />
</div>
`;
